File: api-quarkus/src/main/java/br/ufrn/musi/aplicacao/BuscarObras.java
package br.ufrn.musi.aplicacao;

import br.ufrn.musi.dominio.Dominio.Filtro;
import br.ufrn.musi.dominio.Dominio.Obra;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import java.util.List;

/**
 * `jakarta.enterprise.context` e `jakarta.inject` vêm da especificação CDI, do
 * Jakarta EE. Não são anotações do Quarkus: o mesmo símbolo é reconhecido por
 * qualquer implementação de CDI.
 *
 * Diferenças para Kotlin:
 *
 *   Quarkus + CDI  → o grafo é resolvido em tempo de compilação. Uma dependência
 *                    sem produtor faz o build falhar.
 *   Ktor + Koin    → o grafo é resolvido em tempo de execução, e um `single`
 *                    faltando aparece na inicialização.
 *
 * Nenhuma das duas é errada: uma troca verificação por explicitude, a outra o
 * contrário. Ver docs/decisoes/0001-stacks-e-estrutura.md.
 */
@ApplicationScoped
public class BuscarObras {

    private final FonteDeObras fonte;

    @Inject
    public BuscarObras(FonteDeObras fonte) {   // injeção por construtor
        this.fonte = fonte;
    }

    public List<Obra> executar(Filtro filtro) {
        return fonte.buscar(filtro);
    }
}

File: api-ktor/src/main/kotlin/br/ufrn/musi/aplicacao/BuscarObras.kt
package br.ufrn.musi.aplicacao

import br.ufrn.musi.dominio.Filtro
import br.ufrn.musi.dominio.Obra

/**
 * Caso de uso. Kotlin puro: nenhuma anotação, nenhum framework.
 *
 * A ligação com a implementação é declarada no módulo de Koin, em `Modulos.kt`.
 * O grafo de dependências fica num arquivo só, legível de cima a baixo, em vez de
 * espalhado por anotações.
 *
 * Consequência: esta classe é instanciável com `BuscarObras(fonte)` num teste, sem
 * nenhuma infraestrutura.
 */
class BuscarObras(private val fonte: FonteDeObras) {
    suspend operator fun invoke(filtro: Filtro): List<Obra> = fonte.buscar(filtro)
}
